File: Other/Review/Phase 5/Review1/monotonic-queue/monotonic-queue.go
package main

import (
	"fmt"
)

// MonotonicQueue represents a queue that maintains a monotonic decreasing order
type MonotonicQueue struct {
	queue []int
}

// NewMonotonicQueue is a constructor for MonotonicQueue
func NewMonotonicQueue() *MonotonicQueue {
	return &MonotonicQueue{queue: []int{}}
}

// Push adds a value to the queue while maintaining a monotonically decreasing order
func (mq *MonotonicQueue) Push(value int) {
	for len(mq.queue) > 0 && mq.queue[len(mq.queue)-1] < value {
		mq.queue = mq.queue[:len(mq.queue)-1] // Pop the last element
	}
	mq.queue = append(mq.queue, value)
}

// Pop removes the first element from the queue (FIFO)
func (mq *MonotonicQueue) Pop() {
	if len(mq.queue) > 0 {
		mq.queue = mq.queue[1:]
	}
}

// GetMax returns the maximum element (the first one in the queue)
func (mq *MonotonicQueue) GetMax() *int {
	if len(mq.queue) > 0 {
		return &mq.queue[0]
	}
	return nil
}

// Size returns the number of elements in the queue
func (mq *MonotonicQueue) Size() int {
	return len(mq.queue)
}

// IsEmpty checks if the queue is empty
func (mq *MonotonicQueue) IsEmpty() bool {
	return len(mq.queue) == 0
}

func main() {
	mq := NewMonotonicQueue()
	mq.Push(3)
	mq.Push(1)
	mq.Push(4)
	mq.Push(2)
	mq.Push(1)

	// Print the queue
	fmt.Println("Queue:", mq.queue)

	// Test GetMax
	if max := mq.GetMax(); max != nil {
		fmt.Println("Max:", *max)
	} else {
		fmt.Println("Queue is empty.")
	}

	// Test Pop
	mq.Pop()
	fmt.Println("Queue after Pop:", mq.queue)

	// Test Size and IsEmpty
	fmt.Println("Queue size:", mq.Size())
	fmt.Println("Is Queue empty?", mq.IsEmpty())
}

File: Other/Review/Phase 8/Review1/Part2/monotonic-queue/monotonic-queue.ts
// Monotonically decreasing [4, 2, 1]
export class MonotonicQueue {
  queue: number[];

  constructor() {
    this.queue = [];
  }

  push(value: number) {
    while (this.queue.length && this.queue[this.queue.length - 1] < value) {
      this.queue.pop();
    }
    this.queue.push(value);
  }

  pop() {
    if (this.queue.length) {
      this.queue.shift();
    }
  }

  getMax(): number | undefined {
    return this.queue.length ? this.queue[0] : undefined;
  }

  size(): number {
    return this.queue.length;
  }

  isEmpty(): boolean {
    return this.queue.length === 0;
  }
}

// test
const mq = new MonotonicQueue();
mq.push(3);
mq.push(1);
mq.push(4);
mq.push(2);
mq.push(1);

console.log(mq.queue);
